### ISSUE TITLE
SMP on ESP32 seems to be broken

### ISSUE THREAD

**Issue by alexpaschoaletto** [zephyrproject-rtos/zephyr#83168]
**Describe the bug**
The classic ESP32 board (`esp_wrover_kit/esp32`), one of the few targets I know for having SMP capabilities in Zephyr, is failing to build when `CONFIG_SMP` option is set to `y`.

**To Reproduce**
There's no need to actually try to use any API from SMP, just trying to build should be enough. Head to the `hello_world` sample and insert the following in `prj.conf`:

```c
CONFIG_ESP32_USE_UNSUPPORTED_REVISION=y
CONFIG_SMP=y
```

Now build it for either classic ESP32 variant:
```c
west build samples/hello_world -b esp_wrover_kit/esp32/procpu --pristine
//or
west build samples/hello_world -b esp_wrover_kit/esp32/appcpu --pristine
```

This is what shows up when building for `procpu`:
```cmd
C:/zephyr/soc/espressif/esp32/esp32-mp.c: In function 'esp_appcpu_start':
C:/zephyr/soc/espressif/esp32/esp32-mp.c:184:9: error: 'RTC_CNTL_SW_CPU_STALL' undeclared (first use in this function); did you mean 'RTC_CNTL_SW_CPU_STALL_REG'?
  184 |         RTC_CNTL_SW_CPU_STALL &= ~RTC_CNTL_SW_STALL_APPCPU_C1;
      |         ^~~~~~~~~~~~~~~~~~~~~
      |         RTC_CNTL_SW_CPU_STALL_REG
C:/zephyr/soc/espressif/esp32/esp32-mp.c:184:9: note: each undeclared identifier is reported only once for each function it appears in
C:/Users/axelp/Documents/ISEP/BOLSA-Bolsa_de_Investigacao/6_Github/zephyr/soc/espressif/esp32/esp32-mp.c:185:9: error: 'RTC_CNTL_OPTIONS0' undeclared (first use in this function); did you mean 'RTC_CNTL_OPTIONS0_REG'?
  185 |         RTC_CNTL_OPTIONS0     &= ~RTC_CNTL_SW_STALL_APPCPU_C0;
      |         ^~~~~~~~~~~~~~~~~
      |         RTC_CNTL_OPTIONS0_REG
C:/zephyr/soc/espressif/esp32/esp32-mp.c:186:9: error: 'DPORT_APPCPU_CTRL_B' undeclared (first use in this function); did you mean 'DPORT_APPCPU_CTRL_B_REG'?
  186 |         DPORT_APPCPU_CTRL_B   |= DPORT_APPCPU_CLKGATE_EN;
      |         ^~~~~~~~~~~~~~~~~~~
      |         DPORT_APPCPU_CTRL_B_REG
C:/zephyr/soc/espressif/esp32/esp32-mp.c:187:9: error: 'DPORT_APPCPU_CTRL_C' undeclared (first use in this function); did you mean 'DPORT_APPCPU_CTRL_C_REG'?
  187 |         DPORT_APPCPU_CTRL_C   &= ~DPORT_APPCPU_RUNSTALL;
      |         ^~~~~~~~~~~~~~~~~~~
      |         DPORT_APPCPU_CTRL_C_REG
C:/zephyr/soc/espressif/esp32/esp32-mp.c:190:9: error: 'DPORT_APPCPU_CTRL_A' undeclared (first use in this function); did you mean 'DPORT_APPCPU_CTRL_A_REG'?
  190 |         DPORT_APPCPU_CTRL_A |= DPORT_APPCPU_RESETTING;
      |         ^~~~~~~~~~~~~~~~~~~
      |         DPORT_APPCPU_CTRL_A_REG
```

and this is what happens when you use `appcpu` instead:

```cmd
c:/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd.exe: zephyr/kernel/libkernel.a(sched.c.obj):(.literal.z_thread_halt+0x0): undefined reference to `arch_sched_directed_ipi'
c:/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd.exe: zephyr/kernel/libkernel.a(sched.c.obj): in function `z_thread_halt':
C:/zephyr/kernel/sched.c:457: undefined reference to `arch_sched_directed_ipi'
c:/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd.exe: zephyr/kernel/libkernel.a(smp.c.obj):(.literal.start_cpu+0x8): undefined reference to `arch_cpu_start'
c:/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd.exe: zephyr/kernel/libkernel.a(smp.c.obj): in function `start_cpu':
C:/zephyr/kernel/smp.c:159: undefined reference to `arch_cpu_start'
c:/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd.exe: zephyr/kernel/libkernel.a(ipi.c.obj): in function `signal_pending_ipi':
C:/zephyr/kernel/ipi.c:85: undefined reference to `arch_sched_directed_ipi'
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Expected behavior**
The code should build.

**Impact**
I can't test any SMP application on ESP32 if it doesn't build. I know the board is now "unsupported", so I don't expect any fixes for this particular target. However, if you're dropping support for one of the few boards that feature SMP could you maybe extend this functionality for other targets (e.g. ESP32s3)?

**Environment (please complete the following information):**

 - OS: Windows 11 23H2
 - Toolchain Zephyr 4.0.99, SDK 0.17.0, `west update` executed yesterday.
 - Commit SHA: head is at "ce50ea2f1fb lext: disable for Harvard ARC variants"

**Comment by alexpaschoaletto** [comment]
P.S.: it seems like #56011 and #29394 already state that SMP is not working properly, even when it builds. But this time around the thing won't even build. I'll add to that discussion instead of keeping a separate issue here.
